ENH: add cyclic GC support to object arrays. Closes #6581.

J-Sand opened the pull request:
There are a few things I'm uncertain about here:

- Currently `ndarray` has custom `tp_alloc` and `tp_free` functions, but they don't seem to do anything different to the defaults. Combining cyclic garbage support with custom `tp_alloc` and `tp_free` functions is apparently difficult and I couldn't get it to work, so I just got rid of them. Is that acceptable? It doesn't cause any harm as far as I can tell.

- I don't really understand `UPDATEIFCOPY`. With the more normal use of `base`, the appropriate thing to do is to visit it in `array_traverse` (because `base` can participate in reference cycles) but ignore it in `array_clear` (because you can't have a reference cycle purely consisting of `base`s, AFAIK). I have a feeling we might need to do something special for arrays with `UPDATEIFCOPY` set.

- the `array_traverse` code is largely copied from `PyArray_INCREF`, and I included the code that deals with discontiguous and misaligned arrays. However, `PyArray_INCREF` and friends need to support arrays that don't own their own data, but `array_traverse` (or at least this section of it) doesn't. Is it possible to have a discontiguous or misaligned object array (or array with object fields) that owns its data? If not, `array_traverse` can be simplified quite a bit.

- there are couple of places in `array_dealloc` where the array's refcount is incremented to prevent numpy API calls from triggering `array_dealloc` again (which would be catastrophic). This is harmless, but it makes it difficult to debug GC stuff, because it makes the total refcount go up every time an object array (or an array with `UPDATEIFCOPY` set) is created and then collected. Is there a straightforward way of avoiding this? I can't think of anything.

J-Sand committed:
ENH: add cyclic GC support to object arrays and iterators.

Closes #6581.

J-Sand commented:
OK, I realised my implementation of `array_clear` isn't quite correct. The python docs warn that object pointers have to be set to zero in `tp_clear` before the objects are DECREF'd. I assumed this was in case DECREFing these objects ultimately causes the array's own deallocator to fire, so I wrapped the DECREFs in `Py_INCREF(self); .... Py_DECREF(self);` to prevent that. However, python's GC code already does that for you. The real problem is that the contained objects' deallocators might directly do something to the array themselves. One example where this would happen is if an array contains an array that is an `UPDATEIFCOPY` copy of itself. A more reasonable example would be a custom C extension type.

So, somehow we need to set `data` to zero in `tp_clear` before we start DECREFing any of the pointers. As far as I can see, the easiest way of doing that is to work out how many objects are in the array, allocate a plain C array of that size, copy all of the object pointers into there, dealloc and zero `data`, and then finally walk through the C array and `DECREF` all of the objects. Is there a simpler option? I don't think simply copying the `ndarray` object is an option, in case it's a subclass that does something weird when you copy it (it could maybe even create new ref cycles).

This also leaves us with the problem of what should happen if the deallocator of an array with `UPDATEIFCOPY` fires and it turns out that `base->data` has already been cleared by its own `tp_clear` function. Could we just document that arrays with `UPDATEIFCOPY` set are not guaranteed to update the base array if they somehow end up in a reference cycle?

An example of an `UPDATEIFCOPY` ref cycle is as follows (maybe there is a simpler way of doing it?):

    a = np.zeros(1, dtype=([('a', 'f8'), ('b', 'O')]))
    a['b'][0] = []
    i = np.nditer(a, ['refs_ok'],
            [['readwrite', 'updateifcopy']],
            casting='same_kind',
            op_dtypes=[np.dtype([('a', 'f4'), ('b', 'O')])])
    a['b'][0].append(i.operands[0])
    assert a['b'][0][0].base is a
    assert a['b'][0][0].flags['UPDATEIFCOPY']

Also, I suppose iterator objects are going to need some GC code adding too. Are there any other C-implemented types that can participate in ref cycles?

njsmith commented:
Can you just do

for (int i = 0; i < size; ++i) {
    PyObject* tmp = arr[i];
    arr[i] = NULL;
    DECREF(tmp);
}

? Or is the concern that the DECREF might run arbitrary code, including
writing back to the array's backing store and undoing some of that
clearing? What does PyList do here?

I wouldn't worry too much about UPDATEIFCOPY in cycles -- if an
UPDATEIFCOPY array is in a GC cycle then that's already a bug for sure. The
UPDATEIFCOPY API very much forces the user to make sure that they know
exactly when the UPDATEIFCOPY array is deallocated. There's even a proposal
to deprecate UPDATEIFCOPY in favor of an API that makes this explicit --
it's the last thing needed to get numpy-on-pypy passing the test suite. See
https://mail.scipy.org/pipermail/numpy-discussion/2016-November/076287.html

On Nov 25, 2016 7:23 AM, "J-Sand" <notifications@github.com> wrote:

> OK, I realised my implementation of array_clear isn't quite correct. The
> python docs warn that object pointers have to be set to zero in tp_clear
> before the objects are DECREF'd. I assumed this was in case DECREFing these
> objects ultimately causes the array's own deallocator to fire, so I wrapped
> the DECREFs in Py_INCREF(self); .... Py_DECREF(self); to prevent that.
> However, python's GC code already does that for you. The real problem is
> that the contained objects' deallocators might directly do something to the
> array themselves. One example where this would happen is if an array
> contains an array that is an UPDATEIFCOPY copy of itself. A more
> reasonable example would be a custom C extension type.
>
> So, somehow we need to set data to zero in tp_clear before we start
> DECREFing any of the pointers. As far as I can see, the easiest way of
> doing that is to work out how many objects are in the array, allocate a
> plain C array of that size, copy all of the object pointers into there,
> dealloc and zero data, and then finally walk through the C array and
> DECREF all of the objects. Is there a simpler option? I don't think
> simply copying the ndarray object is an option, in case it's a subclass
> that does something weird when you copy it (it could maybe even create new
> ref cycles).
>
> This also leaves us with the problem of what should happen if the
> deallocator of an array with UPDATEIFCOPY fires and it turns out that
> base->data has already been cleared by its own tp_clear function. Could
> we just document that arrays with UPDATEIFCOPY set are not guaranteed to
> update the base array if they somehow end up in a reference cycle?
>
> An example of an UPDATEIFCOPY ref cycle is as follows (maybe there is a
> simpler way of doing it?):
>
> a = np.zeros(1, dtype=([('a', 'f8'), ('b', 'O')]))
> a['b'][0] = []
> i = np.nditer(a, ['refs_ok'],
>         [['readwrite', 'updateifcopy']],
>         casting='same_kind',
>         op_dtypes=[np.dtype([('a', 'f4'), ('b', 'O')])])
> a['b'][0].append(i.operands[0])
> assert a['b'][0][0].base is a
> assert a['b'][0][0].flags['UPDATEIFCOPY']
>
> Also, I suppose iterator objects are going to need some GC code adding
> too. Are there any other C-implemented types that can participate in ref
> cycles?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/8303#issuecomment-262978384>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAlOaHSF_Zh4T362Ud1EhDt-dJXuB3_lks5rBv1qgaJpZM4K6zXp>
> .
>

J-Sand commented:
`list_clear` copies the number of items and the pointer to the first item into temps, sets them both to zero, and then XDECREFs everything (there is a comment pointing out that the XDECREFs could potentially cause the list to grow again). The problem with that approach is that I can't see a straightforward way of using the iteration API to step through the array once we have copied and zeroed `data`.

I think you're right, it would be OK to leave `data` alone at first, but zero the individual object pointers just before you DECREF them (or maybe set them to `Py_None` just in case anything tries to use them without checking for null pointers?). I'll try that approach. Do you know if I can assume that arrays that own their data are contiguous and aligned? I haven't come across a way of creating one that isn't.

J-Sand committed:
Kill warnings.

J-Sand commented:
Sorry about the noise. I've fixed `array_clear` and several other problems I found, and added support for the iterator objects. I'm fairly confident it's working correctly, but I still have a few concerns:

* I still have code in `array_traverse` and `array_clear` specifically for misaligned/discontiguous object arrays that own their data, and I'm not sure this is possible. Nothing in the test suite reaches this code.

* The way I am dealing with non-object arrays is slightly stupid and inefficient, but I'm not sure there is a better alternative. With the `Py_TPFLAGS_HAVE_GC` flag set, the default `tp_alloc` automatically allocates a little extra memory for each object (two pointers and a `ssize_t` to be precise) for the GC to use, and automatically adds the object to the GC tracker. Everywhere `tp_alloc` is called I have added code to immediately remove the object from the tracker if it can't participate in ref cycles. Ideally `tp_alloc` would automatically decide whether to use the GC allocator or not, but it doesn't receive any information apart from the type object. It might be possible to bypass `tp_alloc` in places like `PyArray_NewFromDescr_int` (at least for `ndarray` itself, probably not for subtypes in case they want to have a custom `tp_alloc`?), but then I think the object would need to be given an extra flag to tell `tp_free` how it was allocated.

* The changes seem to have a big impact on the benchmarks, with some getting much faster and some much slower. I don't know how big an issue that is? I'm guessing it's partly because of the memory layout changes and partly because python's GC routines will be firing at different times.

njsmith commented:
> I still have code in array_traverse and array_clear specifically for misaligned/discontiguous object arrays that own their data, and I'm not sure this is possible. Nothing in the test suite reaches this code.

I think there is one (and only one) way that this can happen, which is that someone calls `PyArray_New` with an explicitly passed in data buffer, the `OWNDATA` flag set, and something like `strides=(16,)` (discontiguous) or `strides=(17,)` (discontiguous and misaligned). Pretty weird case, but I guess technically supported. You could add a test for it.

> Ideally tp_alloc would automatically decide whether to use the GC allocator or not, but it doesn't receive any information apart from the type object.

I guess we *could* smuggle this information in using a thread-local... it's ugly, but depending on how much overhead the GC stuff adds it might be worth it.

> The changes seem to have a big impact on the benchmarks, with some getting much faster and some much slower. I don't know how big an issue that is?

It's definitely a concern -- numpy-using code in general is fairly sensitive to the speed of small array operations, and it would be unfortunate if GC support creates slowdowns on non-object arrays. It's hard to evaluate "some getting much faster and some much slower" though :-). I guess it would help to provide some examples -- maybe using [pyperf](https://pypi.python.org/pypi/perf/0.9.1) to get more reliable numbers on microbenchmarks, plus a few larger/more realistic bits of code just to get a sense how much the overall impact is?

J-Sand commented:
Thanks for the help!

> It's hard to evaluate "some getting much faster and some much slower" though :-)

I ran the benchmarks with something like `run_tests.py --bench-compare HEAD~2 bench_core`, but now I realise it gives me wildly different results every time. Maybe I need some command-line options to increase the number of loops, or it's intended to be run in environments more stable than my laptop?

I ran some benchmarks with `perf` on `np.empty` with dtypes of `float` and `object` and shapes of `1` and `100`. The impact seems pretty negligible:

    before:
    small_empty_float_array: Median +- std dev: 7.21 us +- 0.30 us
    large_empty_float_array: Median +- std dev: 7.13 us +- 0.32 us
    small_empty_object_array: Median +- std dev: 11.8 us +- 0.5 us
    large_empty_object_array: Median +- std dev: 12.8 us +- 0.7 us

    after:
    small_empty_float_array: Median +- std dev: 7.24 us +- 0.26 us
    large_empty_float_array: Median +- std dev: 7.18 us +- 0.30 us
    small_empty_object_array: Median +- std dev: 11.9 us +- 0.5 us
    large_empty_object_array: Median +- std dev: 13.0 us +- 0.6 us

There really shouldn't be much CPU overhead, because tracking/untracking objects just consists of adding/removing them to a doubly linked list. There is no ongoing overhead for untracked objects, and even for object arrays I don't think the traversal should be hugely expensive.

I think the memory is a bigger problem - allocating 24 extra bytes (at least on 64-bit linux) for every array object, which almost always goes unused, is clearly not ideal.

> I guess we could smuggle this information in using a thread-local... it's ugly, but depending on how much overhead the GC stuff adds it might be worth it.

Hmm... the biggest problem is that `tp_free` needs to know which free function to call. We can just have `tp_free` do a `PyDataType_REFCHK`, but then we need to be able to guarantee that all arrays were allocated using the appropriate allocator, and third-party C extension code might fail to do that (for example, somebody might have a subtype with a custom `tp_new`, or they might build an `ndarray` by calling its `tp_alloc` directly, bypassing our `tp_new`).

I think a new flag would work. We could have `tp_alloc` use the non-GC allocator, and everywhere we call `tp_alloc` we could check whether we want an object array, and if so use the GC allocator instead of `tp_alloc` and set the flag. Then `tp_free` could just look at the flag. If an extension module fails to do this dance in the places where they call `tp_alloc`, they won't get an invalid free, they just won't get cyclic GC support (or we could make the GC allocator the default, and they would waste a bit of memory). And if they already bypass our `tp_alloc` altogether, they really should be providing their own `tp_free`, right?

That's the best option I can think of anyway.

J-Sand commented:
So, here's a more interesting benchmark:

    class A: pass
    a = np.empty(10000, dtype='O')
    for j in range(10000):
        a[j] = np.empty(1000, dtype='O')

We're allocating lots of (GC-aware) objects and keeping them in memory, which means that the GC is doing lots of traversals, and also we have lots of big object arrays, so traversals are expensive. With my changes it slows down by about a factor of 2.5. I think this should be just about the worst-case scenario, and I can't find a benchmark seeing a significant slowdown that doesn't involve keeping lots of (or very large) object arrays in memory.

So I think the question is whether avoiding memory leaks caused by object arrays in cycles is worth the extra memory for array objects, plus slowdowns in cases like this. The only way I can see to get around the memory issue is to give object arrays their own subtype, which is exactly like `ndarray` except that it has GC support. Though I guess that would have severe backwards compatibility implications.

homu commented:
:umbrella: The latest upstream changes (presumably #8903) made this pull request unmergeable. Please resolve the merge conflicts.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1753,22 +1764,258 @@ array_iter(PyArrayObject *arr)
+            if (!PyArg_ParseTuple(value, "Oi|O", &new, &offset,
+                                  &title)) {
+                PyErr_Print();
+                PyErr_Clear();
```
Are you looking for `PyErr_WriteUnraiseable` here?

seberg commented:
Should we resurrect this? There are some smaller things to clean up, and the (non-trivial) traverse/clear paths can go into `refcount.c` and making that `refcount.c.src` and reduce the code.

seberg commented:
Since I think the code was already very far, I am looking into it. Will be some minor fix-ups mostly.

eric-wieser commented:
@seberg: Did you get anywhere with your local changes to this?

seberg commented:
Hmmm, yeah, I have a local branch based on this. Somehow lost track of it. It was probably fairly far along, although I am sure I will have some merge conflicts and there must have been some reason I had not created a PR yet.

seberg commented:
Closing this in favor of gh-15065 which resolves the merge conflict and fixes it up more generally.
